File: sources/android/app/src/main/java/com/siarheikuchuk/ftpsserver/ui/ConnectionDetails.kt
package com.siarheikuchuk.ftpsserver.ui

import android.content.Context
import com.siarheikuchuk.ftpsserver.R

fun buildConnectionDetails(context: Context, state: UiState): String {
    fun s(id: Int) = context.getString(id)
    val hosts = buildList {
        for (net in state.networks) {
            addAll(net.addresses)
        }
        if (state.hostName.isNotBlank()) add(state.hostName)
    }.distinct()

    val b = StringBuilder()
    fun line(text: String? = null) {
        if (text != null) b.append(text)
        b.append('\n')
    }

    line(s(R.string.connection_details_title))
    line()
    line(s(R.string.connection_details_intro_android))
    line()
    line(s(R.string.connection_details_clients))
    line()
    line(s(R.string.connection_details_fill_fields))
    line()
    line(s(R.string.connection_details_host_title))
    line()
    line(s(R.string.connection_details_host_body))
    line()
    for (host in hosts) line("  $host")
    line()
    line(s(R.string.connection_details_port_title))
    line()
    line(s(R.string.connection_details_port_body))
    line()
    line("  ${state.port}")
    line()
    line(s(R.string.connection_details_encryption_title))
    line()
    line(s(R.string.connection_details_encryption_body))
    line()
    line(s(R.string.connection_details_accounts_title))
    line()
    line(s(R.string.connection_details_accounts_body))
    line()
    state.users.forEachIndexed { index, user ->
        val letter = if (index < 26) ('a' + index).toString() else (index + 1).toString()
        val access = s(if (user.readonly) R.string.connection_details_access_read_only else R.string.connection_details_access_read_write)
        val folder = user.folderName.ifBlank { s(R.string.connection_details_folder_unspecified) }
        line(
            "  " + context.getString(
                R.string.connection_details_account_format,
                letter,
                user.login,
                access,
                folder,
            ),
        )
        line()
        line("     ${s(R.string.connection_details_login_label)}     ${user.login}")
        line("     ${s(R.string.connection_details_password_label)}  ${user.password}")
        line()
    }

    val cert = state.certificate
    if (state.useSelfSigned && cert != null && cert.isSelfSigned) {
        line(s(R.string.connection_details_cert_title))
        line()
        line(s(R.string.connection_details_cert_body))
        line()
        line("  " + context.getString(R.string.connection_details_cert_sha256_format, cert.fingerprintSha256))
        line()
        line("  " + context.getString(R.string.connection_details_cert_sha1_format, cert.fingerprintSha1))
        line()
        line(s(R.string.connection_details_cert_fingerprint_hint))
    }

    return b.toString().trimEnd() + "\n"
}

File: sources/FtpsServerWindows/Helpers/ConnectionDetails.cs
using FtpsServerAppsShared.Helpers;
using FtpsServerWindows.Models;
using FtpsServerWindows.Resources;

namespace FtpsServerWindows.Helpers;

internal static class ConnectionDetails
{
    public static string Build(
        int port,
        IEnumerable<UserAccount> users,
        bool selfSigned,
        string? sha256,
        string? sha1)
    {
        return ConnectionDetailsText.Build(
            new ConnectionDetailsInput
            {
                Hosts = ConnectionDetailsText.CollectHostValues(Environment.MachineName),
                Port = port.ToString(),
                Accounts = users.Select(user => new ConnectionDetailsAccount
                {
                    Login = user.Login,
                    Password = user.Password,
                    Folder = user.Folder,
                    ReadOnly = user.ReadonlyPermission,
                }).ToList(),
                IsSelfSigned = selfSigned,
                Sha256 = sha256,
                Sha1 = sha1,
            },
            CreateStrings());
    }

    public static ConnectionDetailsStrings CreateStrings() => new()
    {
        Title = Strings.ConnectionDetailsTitle,
        Intro = Strings.ConnectionDetailsIntroDesktop,
        Clients = Strings.ConnectionDetailsClients,
        FillFields = Strings.ConnectionDetailsFillFields,
        HostTitle = Strings.ConnectionDetailsHostTitle,
        HostBody = Strings.ConnectionDetailsHostBody,
        PortTitle = Strings.ConnectionDetailsPortTitle,
        PortBody = Strings.ConnectionDetailsPortBody,
        EncryptionTitle = Strings.ConnectionDetailsEncryptionTitle,
        EncryptionBody = Strings.ConnectionDetailsEncryptionBody,
        AccountsTitle = Strings.ConnectionDetailsAccountsTitle,
        AccountsBody = Strings.ConnectionDetailsAccountsBody,
        AccountFormat = Strings.ConnectionDetailsAccountFormat,
        LoginLabel = Strings.ConnectionDetailsLoginLabel,
        PasswordLabel = Strings.ConnectionDetailsPasswordLabel,
        AccessReadWrite = Strings.ConnectionDetailsAccessReadWrite,
        AccessReadOnly = Strings.ConnectionDetailsAccessReadOnly,
        FolderUnspecified = Strings.ConnectionDetailsFolderUnspecified,
        CertTitle = Strings.ConnectionDetailsCertTitle,
        CertBody = Strings.ConnectionDetailsCertBody,
        CertSha256Format = Strings.ConnectionDetailsCertSha256Format,
        CertSha1Format = Strings.ConnectionDetailsCertSha1Format,
        CertFingerprintHint = Strings.ConnectionDetailsCertFingerprintHint,
    };
}
